xilem/examples/stopwatch.rs

```diff
@@ -0,0 +1,260 @@
+    // The duration displayed; updated by by `update_display()`
+    displayed_duration: Duration,
+    // An error string to display if there is an error.
+    displayed_error: String,
+    // A list of the length of all completed splits. Does not include the current split.
+    completed_lap_splits: Vec<Duration>,
+    // The duration of the main timer when the split was started.
+    split_start_time: Duration,
+}
+
+impl Stopwatch {
+    fn start(&mut self) {
+        self.last_start_time = Some(SystemTime::now());
+        self.active = true;
+        self.update_display();
+    }
+
+    fn stop(&mut self) {
+        let dur_since_last_instant = self
+            .last_start_time
+            .expect("stop should only be called when the start time is set")
+            .elapsed();
+        match dur_since_last_instant {
+            Ok(dur) => {
+                self.added_duration = self.added_duration.add(dur);
+                self.displayed_error = "".into();
+            }
+            Err(err) => {
+                self.displayed_error = format!("failed to calculate elapsed time: {}", err);
+            }
+        }
+        self.last_start_time = None;
+        self.active = false;
+        self.update_display();
+    }
+
+    fn reset(&mut self) {
+        self.active = false;
+        self.last_start_time = None;
+        self.added_duration = Duration::ZERO;
+        self.completed_lap_splits.clear();
+        self.split_start_time = Duration::ZERO;
+        self.update_display();
+    }
+
+    fn lap(&mut self) {
+        let split_end = self.get_current_duration();
+        self.completed_lap_splits
+            .push(split_end.sub(self.split_start_time));
+        self.split_start_time = split_end;
+    }
+
+    fn get_current_duration(&self) -> Duration {
+        match self.last_start_time {
+            Some(last_start_time) => match last_start_time.elapsed() {
+                Ok(elapsed) => self.added_duration + elapsed,
+                Err(err) => {
+                    warn!("error calculating elapsed time: {}", err.to_string());
+                    self.added_duration
+                }
+            },
+            _ => self.added_duration,
+        }
+    }
+
+    fn update_display(&mut self) {
+        self.displayed_duration = self.get_current_duration();
+    }
+}
+
+fn get_formatted_duration(dur: &Duration) -> String {
+    let seconds = dur.as_secs_f64() % 60.0;
+    let minutes = (dur.as_secs() / 60) % 60;
+    let hours = (dur.as_secs() / 60) / 60;
+    format!("{hours}:{minutes:0>2}:{seconds:0>4.1}")
+}
+
+fn app_logic(data: &mut Stopwatch) -> impl WidgetView<Stopwatch> {
+    fork(
+        flex((
+            FlexSpacer::Fixed(5.0),
+            label(get_formatted_duration(&data.displayed_duration)).text_size(70.0),
+            flex((lap_reset_button(data), start_stop_button(data))).direction(Axis::Horizontal),
+            FlexSpacer::Fixed(1.0),
+            laps_section(data),
```

Comment:
I wonder whether there should be a "legend-row" above this, explaining the two different times?

Reply:
That would really benefit from a future grid layout type. Getting the alignment right with the current options would be a little fragile.